internal/guest/storage/crypt/utilities.go

```diff
@@ -28,123 +28,52 @@ func getBlockDeviceSize(ctx context.Context, path string) (int64, error) {
 
 	pos, err := file.Seek(0, io.SeekEnd)
 	if err != nil {
-		return 0, errors.Wrap(err, "error seeking end of: "+path)
+		return 0, fmt.Errorf("error seeking end of %s: %w", path, err)
 	}
-
 	return pos, nil
 }
 
-// createSparseEmptyFile creates a sparse file of the specified size. The whole
-// file is empty, so the size on disk is zero, only the logical size is the
-// specified one.
-func createSparseEmptyFile(ctx context.Context, path string, size int64) (err error) {
-	f, err := os.Create(path)
+// In the xfs mkfs case it appears to attempt to read the first block of the device.
+// This results in an integrity error. This function zeros out the start of the device
+// so we are sure that when it is read it has already been hashed so matches.
+func zeroDevice(devicePath string, blockSize int64, numberOfBlocks int64) error {
+	fout, err := os.OpenFile(devicePath, os.O_WRONLY, 0)
 	if err != nil {
-		return errors.Wrapf(err, "failed to create: %s", path)
+		return fmt.Errorf("failed to open device file %s: %w", devicePath, err)
 	}
+	defer fout.Close()
 
-	defer func() {
-		if err != nil {
-			if inErr := os.RemoveAll(path); inErr != nil {
-				log.G(ctx).WithError(inErr).Debug("failed to delete: " + path)
-			}
-		}
-	}()
-
-	defer func() {
-		if err := f.Close(); err != nil {
-			log.G(ctx).WithError(err).Debug("failed to close: " + path)
-		}
-	}()
-
-	if err := f.Truncate(size); err != nil {
-		return errors.Wrapf(err, "failed to truncate: %s", path)
-	}
-
-	return nil
-}
-
-// The following constants aren't defined in the io or os libraries.
-//
-//nolint:stylecheck // ST1003: ALL_CAPS
-const (
-	SEEK_DATA = 3
-	SEEK_HOLE = 4
-)
-
-// copyEmptySparseFilesystem copies data chunks of a sparse source file into a
-// destination file. It skips holes. Note that this is intended to copy a
-// filesystem that has just been generated, so it only contains metadata blocks.
-// Because of that, the source file must end with a hole. If it ends with data,
-// the last chunk of data won't be copied.
-func copyEmptySparseFilesystem(source string, destination string) error {
-	fin, err := os.OpenFile(source, os.O_RDONLY, 0)
-	if err != nil {
-		return errors.Wrap(err, "failed to open source file")
+	zeros := make([]byte, blockSize)
+	for i := range zeros {
+		zeros[i] = 0
```

Comment:
Doenst go already zero-out slices on creation?

Reply:
I'd have guessed so but I could not find it written down as a guarantee.
In fact it could almost be any old rubbish written here as the point is to provoke a write before mkfs.xfs reads sector 0. However, failing to set the contents of the block could leak information contained in that memory.

Reply:
and i guess the compiler should be smart enough to skip this if it is guaranteed to be zero too